### PR TITLE
[provider_azure] Extract "groups" claim from jwt and add it to Session

### PR DESCRIPTION
**Pull request by daddonpa**
## Description

The azure oauth server can provide additional claims in the id token which is a jwt.

In azure, you can configure additional claims for a service principal in general as described here
[https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims).

The `groups` claim is a bit special, as its value comes from an existing value for the user and not a single static value 
[https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims)  

This pull request adds code to
- extract this `groups` claim from the jwt inside the `EnrichSession` method by
  - splitting the token into the 3 jwt parts
  - base64 decoding the center part
  - parsing that part into a json structure
  - extract the value for the key `groups` from it
  - assigns that value to `session.Groups`

## Motivation and Context

- In a project I work for I want to grant access to a specific domain for some groups but not others
- I want to use a single azure oauth server to manage all users and groups that might have access to some parts of my application
- I want to then grant or deny access to certain routes based on the group using `allowed_groups`

This has already been requested by another user here: https://github.com/oauth2-proxy/oauth2-proxy/issues/1040
So I implemented this and deployed and tested it for my use case.

## How Has This Been Tested?

- Added a test `TestAzureProviderGetGroupsFromJWT` in `azure_test.go` to test and debug the jwt parsing
- Built the docker image using from this branch and deployed it into a kubernetes cluster like this
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: oauth2-proxy
  namespace: default
spec:
  template:
    spec:
      containers:
      - args:
        - --http-address=0.0.0.0:4180
        - --email-domain=*
        - --cookie-secret=$(COOKIE_PASSWORD)
        - --reverse-proxy=true
        - --upstream=http://127.0.0.1:8080/
        - --session-store-type=redis
        - --redis-connection-url=redis://redis-master.storage/
        - --redis-password=$(REDIS_PASSWORD)
        - --config=/etc/oauth2_proxy/oauth2_proxy.cfg
```
- Created an ingress to grant access to a grafana based on azure group
```
kind: Ingress
  metadata:
    name: grafana-grafana-oauth2
    namespace: monitoring
    labels:
    annotations:
      kubernetes.io/ingress.class: "monitoring-cluster"
      nginx.ingress.kubernetes.io/auth-signin: "https://grafana.daddona.de/oauth2/sign_in"
      nginx.ingress.kubernetes.io/auth-url: "https://grafana.daddona.de/oauth2/auth?allowed_groups=a,b"
  spec:
    tls:
      - hosts:
        - grafana.daddona.de
        secretName: grafana.daddona.de
    rules:
     - host: grafana.daddona.de
       http:
         paths:
           - path: /
             backend:
               serviceName: grafana
               servicePort: 80
```
- Created azure groups `a`, `b`, and `c`
- Add azure account to group `a` and try that access works
- Add azure account to group `c` but not `a` and try that access is denied 

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/azure.go [hunk 1]
```diff
@@ -283,3 +286,28 @@ func (p *AzureProvider) GetLoginURL(redirectURI, state string) string {
 func (p *AzureProvider) ValidateSession(ctx context.Context, s *sessions.SessionState) bool {
 	return validateToken(ctx, p, s.AccessToken, makeAzureHeader(s.AccessToken))
 }
+
+func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	jwtParts := strings.Split(s.IDToken, ".")
```
Comment (NickMeves) [inline providers/azure.go:291]: This will likely need to wait until this related PR for Azure wraps up: https://github.com/oauth2-proxy/oauth2-proxy/pull/914

You will need to take advantage of what that PR is now using to access the claims.

We don't trust any claims unless you verify the signature (There's a robust discussion in that PR on this stance if you want to take a peek).

There's new helper OIDC methods to facilitate this that PR is using to extract the `email` claim from the `id_token` -- you can likely take advantage of the same codepath.

Comment (daddonpa) [reply providers/azure.go]: That makes sense. That pull request should cover almost everything I intend to do with mine, we just focused on different claims.

I did now pick the token verification and claim extraction from https://github.com/oauth2-proxy/oauth2-proxy/pull/914 into this PR and tested it with `oidc_issuer_url = "https://sts.windows.net/}tenantid}/"` in the config and this works fine.

After the other PR is merged I will check if my changes are still needed and if so rebase them and update this request

**Comment by NickMeves** [comment]
The other PR I mentioned merged. Once you rebase ping me and I'll give this a thorough review.

**Comment by acobaugh** [comment]
I think this is now just a simple matter of doing something like this in a few places, no?

```
		if len(session.Groups) == 0 && len(claims.Groups) > 0 {
			session.Groups = claims.Groups
		}
```

**Comment by daddonpa** [comment]
@acobaugh Yes I plan to implement something similar to the `verifyTokenAndExtractEmail` (https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/azure.go#L221) and call it from `EnrichSession`. Once I tested it I will update the merge request.

**Comment by acobaugh** [comment]
Roger that. Our code base is slightly different, but this is all I did to get groups working: https://github.com/PennState/oauth2-proxy/blob/azure_testing/providers/azure.go#L197

**Comment by daddonpa** [comment]
@NickMeves I rebased the changes to the latest version of master.

I am now using the existing `verifyTokenAndExtractClaims` to get the groups the same way as done for the email.

**Comment by daddonpa** [comment]
Hi @NickMeves ,
could you please have a look at the changes?

File: providers/azure.go [hunk 2]
```diff
@@ -177,21 +177,27 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
...
 
-	email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
-	if err != nil {
-		return fmt.Errorf("unable to get email address: %v", err)
-	}
-	if email == "" {
-		return errors.New("unable to get email address")
+	if s.Groups == nil {
+		// add groups claim
+		oidcClaims, err := p.verifyTokenAndExtractClaims(ctx, s.IDToken)
```
Comment (NickMeves) [inline providers/azure.go:196]: You are verifying & extracting the same token/claims twice it appears (once for email another for groups)?

I think this should be refactored to just handle email & groups (if applicable) from the same extracted claims. OIDCProvider should have samples of this pattern I imagine.

This is also not properly handling any non-nil errors (which likely won't happen because they would happen in the call above for emails 😄 ) -- so unifying will clear up the logic and gaps I think.

Comment (NickMeves) [reply providers/azure.go:196]: Scratch that - I mistook `getEmailFromProfileURL` for `verifyTokenAndExtractEmail` -- but that is also called in the chain, so I think the double call concern stands.

Comment (NickMeves) [reply providers/azure.go:196]: I think you want to emulate the `createSession` helper method in the OIDCProvider that will build the user/email/groups from the claims and is used wherever needed (initial redeem, refreshes).

`EnrichSession` is more intended for network calls to fetch any missing data only in the initial login.

If you moved all the claims logic together in the `createSession` parallel, you'd also get the missing functionality of setting the group from claims in refresh calls (which I believe this is missing at the moment?)

Comment (NickMeves) [reply providers/azure.go:196]: The current `verifyTokenAndExtractEmail` is attached in the proper places, I think you want the `group` claim extraction to happen there.

File: docs/docs/configuration/group.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+You will need an oauth provider that supports adding the `group` claim in the `id_token`.
+
+Configured providers are :
+
+- [Azure](#azure-auth-provider)
+
+### Azure Auth Provider
+
+In azure, you can configure additional claims for a service principal in general as described here
+[https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims).
```
Comment (JoelSpeed) [inline docs/docs/configuration/group.md:16]: Might be better to have an inline link, perhaps we can reword it to flow more naturally to allow a link
```suggestion
Before you can use groups with the Azure provider, you must [configure additional claims](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims) for a service principal using the Azure portal.
```
WDYT?

File: docs/docs/configuration/group.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+The `groups` claim is a bit special, as its value comes from an existing value for the user and not a single static value 
+[https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims)
```
Comment (JoelSpeed) [inline docs/docs/configuration/group.md:19]: Would be good to make this a more natural link too if possible

File: docs/docs/configuration/group.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+[https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims).
+
+The `groups` claim is a bit special, as its value comes from an existing value for the user and not a single static value 
+[https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims](https://docs.microsoft.com/en-us/azure/active-directory/hybrid/how-to-connect-fed-group-claims)  
+
+1. Configure the service principal to get the groups from the logged in user   
+   ```bash
+   az ad app update --id <sp_object_id> --set groupMembershipClaims=ApplicationGroup
+   ```
+   The option here are `All` where every group the user is assigned to is sent to the oauth2-proxy.
```
Comment (JoelSpeed) [inline docs/docs/configuration/group.md:25]: I think this should be `options` right?
```suggestion
   The options here are `All` where every group the user is assigned to is sent to the oauth2-proxy. 
```

File: docs/docs/configuration/group.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  ```go
+  EnrichSession(ctx context.Context, s *sessions.SessionState) error;
+  ```
+  method to extract the `groups` claim from the `id_token` returned by azure
+* [`oidc.go`](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/providers/oidc.go)
+  Using a separate implementation
+  ```go
+  createSessionState(ctx context.Context, token *oauth2.Token, idToken *oidc.IDToken) (*sessions.SessionState, error);
+  ```
+  called from the `Redeem` and `redeemRefreshToken` methods
```
Comment (JoelSpeed) [inline docs/docs/configuration/group.md:85]: I'm not certain whether we would want this kind of documentation in the docs here, this is typically just user docs. I'd suggest we would want a separate section for developer docs like this :thinking:

File: providers/azure.go [hunk 7]
```diff
@@ -177,21 +177,27 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
...
+		s.Email = email
 	}
 
-	email, err := p.getEmailFromProfileAPI(ctx, s.AccessToken)
-	if err != nil {
-		return fmt.Errorf("unable to get email address: %v", err)
-	}
-	if email == "" {
-		return errors.New("unable to get email address")
+	if s.Groups == nil {
```
Comment (JoelSpeed) [inline providers/azure.go:194]: If groups is a list, do we need to check for non nil here or do we just want to check for the length of the list?

File: providers/azure.go [hunk 8]
```diff
@@ -216,18 +222,32 @@ func (p *AzureProvider) prepareRedeem(redirectURL, code string) (url.Values, err
...
+	}
+
+	return email, nil
+}
+
+// verifyTokenAndExtractClaims tries to extract oauth claims from either id_token or access token
+// when oidc verifier is configured
+func (p *AzureProvider) verifyTokenAndExtractClaims(ctx context.Context, token string) (*OIDCClaims, error) {
+	var oidcClaims *OIDCClaims
+	oidcClaims = nil
```
Comment (JoelSpeed) [inline providers/azure.go:242]: The second line here is redundant, initialising a var like this means it is nil already

File: providers/azure.go (hunk 2 above)
Comment (JoelSpeed) [reply providers/azure.go:196]: I think if we were to reorder this a bit, we could avoid calling it twice. For example, check at first if both email and groups are empty, return early, if not, we need the oidcClaims to populate at least one of them, then perform the rest of the logic with only fetching it once

I had intended to make a generic utility for extracting claims like this, a `claimExtractor` kind of thing that would be able to used to get a named claim, and would only do the verification if there was a need to actually get a token, or could also look at the profile URL if needed. Maybe this would be a good time to do that.

**Comment by michel-zimmer** [comment]
Hi everyone, any chance in getting this on track again?
I do not understand much of it, but it looks promising to me.
